README.md
# Deprecated branch

This version of Dicord Accept rules bot is deprecated and will no longer be supported.
Please switch to the master branch and download a current release!

# Discord Accept Rules Module

This is a module for my Discord Bot Mainframe. You can find it [here](https://go.astrogd.eu/mainframe).

## Installation

### Files
- Put the .js File into the modules folder of the mainframe
- Put the .db File into the db folder of the mainframe
- Put the .json File into the config folder of the mainframe

### Config

- Edit config/discord-accept-rules-bot.json and insert the required values:
    - guildid: The ID of the server, the Bot should listen to (Example: 195221422682013696).
    - roleid: The ID of the role, the Bot should give Users, that accepted the rules.
    - embedColor: The Hex Color with the "#" that should be used for the Embed, the Bot sends where you can accept the rules.
    - verifyMessageTitle: The Title of the Embed
    - verifyMessageDescription: The Description of the Embed
    - verifyMessageFields: Field Values if you need more content inside your embed (Scroll down for more information)
    - verifyMessageReaction: Reactioncode of the Reaction that is being used to verify the rules. To get the Reactioncode, search your emoji [here](https://emojipedia.org/) and click on "copy".

### Enable the module in main.js
- Open the main.js file of your mainframe version in a code-editor of your choice
- Under Module Definition paste the following code:

    ```const discord_accept_rules_bot = require(`${__dirname}/modules/discord-accept-rules-bot.js`)(client, fs, tools, __dirname);```
- Under Module Commands **after** ```if (await tools.command(cmd, args)) success = true;``` paste the following code:

    ```if (await discord_accept_rules_bot.command(cmd, args)) success = true;```
- Under Module Initalization inside the Curly-Brackets paste the following code:

    ```await discord_accept_rules_bot.init();```
- The Module should now be enabled.

### verifyMessageFields

verifyMessageFields contains an Array with an object for each additional field. Each Object can contain the following parameters:

- title: (String) The Title of your Field
- value: (String) The content of your Field (supports formatting)
- inline: (Boolean, Default false) true if you want the field to be on the right side of the field above. You can create Table like Structures with this option
- empty: (Boolean, Default false) if true an empty Field wil be created for spacing. title and value are being ignored, inline not!

Example:

   ```
   [
    {
     "title": "Title",
     "value": "Content"
    },
    {
     "title": "Title2",
     "value": "Content2",
     "inline": true
    },{
     "empty": true
    },{
     "title": "After Empty",
     "value": "After Empty content"
    }
   ]
   ```

### Role and Serverids

To get Role and Serverids, you may use a bot, to keep things simple. I recommend [DynoBot](https://www.dynobot.net/).
You can kick Dyno after you've got the required ID's

Get Serverid: ?serverinfo --> On the bottom of the Embed you can find ID: "SOME ID HERE"
Get Role-ID: ?roleinfo <Rolename> --> First field contains the ID of the role
    
### Thank you for using my Bot
If you have any questions feel free to contact me:

Twitter: @astrogd

Instagram: @astrogd.eu

Email: support@astrogd.eu

Discord: AstroGD#3416 [Discord Server](https://www.discord.me/astrogd)



changelog.md
# TBA
#### Released: TBA

## Updates
:arrow_up: Updated follow-redirects to 1.14.0

# V2.2.0
#### Released: 11.04.2021

## New
:memo: Add changelog.md

## Changes
:recycle: Refactor software version checks to simplify update notifications
:memo: Update Readme for the separation of no longer supported mainframe

## Dependency updates
:arrow_up: discord.js to 12.5.3
:arrow_up: follow-redirects to 1.13.3

tools/tools.js
/**
 * Tools js
 * @author AstroGD - https://www.astrogd.eu
 * @since 2018-07-15
 */
var scriptName = "tools.js";
const path = require("path");

module.exports = function (Discord, client, fs, dir) {
    const package = require(path.join(__dirname, "../package.json"));

    async function init() {
        let currdate = new Date();
        this.startupdate = `${currdate.getUTCFullYear()}-${currdate.getUTCMonth() + 1}-${currdate.getUTCDate()}-${currdate.getUTCHours()}-${currdate.getUTCMinutes()}-${currdate.getUTCSeconds()}`;

        await fs.ensureDirSync(`${dir}/log/`);
        this.stream = await fs.createWriteStream(`${dir}/log/${this.startupdate}.log`);

        this.initialized = true;

        this.stream.write(`[${this.startupdate}] INFO [${scriptName}] ${scriptName} V${package.version} initialized\n`);
        console.log(`[${this.startupdate}] INFO [${scriptName}] ${scriptName} V${package.version} initialized`);
    }

    function log(x, msg, state) {
        if (!this.initialized) return;

        let currdate = new Date();
        let timestampCurrent = `${currdate.getUTCFullYear()}-${currdate.getUTCMonth() + 1}-${currdate.getUTCDate()}-${currdate.getUTCHours()}-${currdate.getUTCMinutes()}-${currdate.getUTCSeconds()}`;

        switch (state) {
            case 1: //ERROR
                state = "ERROR";
                break;
            case 2: //WARN
                state = "WARN";
                break;
            default: //INFO
                state = "INFO";
                break;
        }

        let logCurrent = `[${timestampCurrent}] ${state} [${x}] ${msg}`;
        this.stream.write(logCurrent + '\n');
        console.log(logCurrent);
    }

    function logFile(x, msg, state) {
        if (!this.initialized) return;

        let currdate = new Date();
        let timestampCurrent = `${currdate.getUTCFullYear()}-${currdate.getUTCMonth() + 1}-${currdate.getUTCDate()}-${currdate.getUTCHours()}-${currdate.getUTCMinutes()}-${currdate.getUTCSeconds()}`;

        switch (state) {
            case 1: //ERROR
                state = "ERROR";
                break;
            case 2: //WARN
                state = "WARN";
                break;
            default: //INFO
                state = "INFO";
                break;
        }

        let logCurrent = `[${timestampCurrent}] ${state} [${x}] ${msg}`;
        this.stream.write(logCurrent + '\n');
    }

    async function command(cmd, _args) {
        if (!this.initialized) return;
        switch (cmd) {
            case "help":
                console.log(`===== Help Commands from Modules and Mainframe\n\n`);
                break;

            default:
                break;
        }
    }

    function isVersionLower(base, check) {
        base = base.split(".");
        check = check.split(".");

        if (check[0] < base[0]) return true;
        if (check[0] == base[0] && check[1] < base[1]) return true;
        if (check[0] == base[0] && check[1] == base[1] && check[2] < base[2]) return true;

        return false;
    }

    function makeEmbed(author, color, title, description, fields) {
        try {
            var embed = new Discord.MessageEmbed();
            embed.setColor(color)
                .setFooter(`AstroGD Discord Mainframe System V${mainframeversion}`, client.user.avatarURL)
                .setTimestamp(new Date())
                .setTitle(title)
                .setDescription(description);

            if (author) {
                embed.setAuthor(author.username, author.avatarURL);
            }

            if (fields) {
                for (var i = 0; i < fields.length; i++) {
                    if (!fields[i].inline) fields[i].inline = false;
                    if (!fields[i].empty) {
                        embed.addField(fields[i].title, fields[i].value, fields[i].inline);
                    } else {
                        embed.addField('\u200b', '\u200b', fields[i].inline);
                    }
                }
            }
            return embed;
        } catch (e) {
            throw e;
        }
    }

    function getServerVersionPack() {
        return new Promise((resolve, reject) => {
            const https = require('follow-redirects').https;

            var options = {
                'method': 'GET',
                'hostname': 'software.astrogd.eu',
                'path': '/accept-rules/version.json',
                'headers': {},
                'maxRedirects': 20
            };

            var req = https.request(options, function (res) {
                var chunks = [];

                res.on("data", function (chunk) {
                    chunks.push(chunk);
                });

                res.on("end", function () {
                    var body = Buffer.concat(chunks);
                    resolve(body);
                });

                res.on("error", function (error) {
                    reject(error);
                });
            });

            req.end();
        });
    }

    async function checkVersion() {
        let versionInfo = null;
        try {
            versionInfo = await getServerVersionPack();
        } catch (error) {
            this.log(scriptName, `Version couldn't be checked! Make sure to allow network access to software.astrogd.eu`, 1);
            return null;
        }

        if (!versionInfo) {
            this.log(scriptName, `Version couldn't be checked! Make sure to allow network access to software.astrogd.eu`, 1);
            return null;

        }

        try {
            versionInfo = JSON.parse(versionInfo.toString());
        } catch (error) {
            this.log(scriptName, `Something went wrong while checking for a newer version. Please check the repository manually to see if there is a new version available.`, 1);
            return null;
        }

        return versionInfo;
    }

    return {
        init,
        log,
        logFile,
        command,
        makeEmbed,
        checkVersion,
        isVersionLower
    }
}

discord-accept-rules-bot.js
/**
 * Discord Accept Rules Bot
 * @version 1.0.1
 * @author AstroGD - https://www.astrogd.eu
 * @since 2018-07-15
 */

var scriptName = "modules.discord-accept-rules-bot.js",
    version = "1.0.1",
    mainframeMinVersion = "1.0.1";

module.exports = function (client, fs, tools, dir) {

    async function init(deactivated) {
        /**
         * @deprecated This branch is deprecated
         */

        this.initialized = false;
        this.activated = false;

        console.error("This branch is deprecated and no longer supported. Please switch to master branch!");
        process.exit();

        tools.log(scriptName, `Initializing ${scriptName} V${version}`);

        if (tools.checkMainframeVersion(mainframeMinVersion)) {
            this.initialized = true;
        } else {
            tools.log(scriptName, `This Module requires a Mainframe Version of ${mainframeMaxVersion} or higher. Please update your Mainframe. This Module is not being activated.`, 2);
            return;
        }
        if (!deactivated) {
            this.activated = true;
        }

        this.db = JSON.parse(fs.readFileSync(`${dir}/db/discord-accept-rules-bot.db`, "utf-8"));
        this.config = require(`${dir}/config/discord-accept-rules-bot.json`);

        if (!client.guilds.get(this.config.guildid)) {
            tools.log(scriptName, `The guild specified in the modules config was not found. Please check the config file. Verifymessage could not be sent.`, 1);
            this.initialized = false;
            return;
        }

        if (!this.db.messageid) {
            tools.log(scriptName, `There is no Welcome Message saved. Please add one using "addverifymessage <Channelid>" in the command line`, 2);
        } else {
            let guild = client.guilds.get(this.config.guildid);
            if (guild) {
                let channel = guild.channels.get(this.db.messagechannelid);
                if (!channel) {
                    tools.log(scriptName, `There is no Welcome Message saved. Please add one using "addverifymessage <Channelid>" in the command line`, 2);
                } else {
                    try {
                        await channel.fetchMessage(this.db.messageid);
                    } catch (e) {
                        tools.log(scriptName, `There is no Welcome Message saved. Please add one using "addverifymessage <Channelid>" in the command line`, 2);
                    }
                }
            }

        }

        client.on("messageReactionAdd", async (reaction, user) => {
            if (user.bot) return;
            if (reaction.emoji.name == this.config.verifyMessageReaction && reaction.message.id == this.db.messageid) {
                let member = reaction.message.guild.members.get(user.id);
                if (!member) return;

                if (!member.roles.get(this.config.roleid)) {
                    let role = reaction.message.guild.roles.get(this.config.roleid);
                    if (!role) {
                        tools.log(scriptName, `The specified role was not found on the server. Please check modules config.`, 1);
                        return;
                    }

                    try {
                        await member.addRole(role);
                    } catch (e) {
                        tools.log(scriptName, `Couldn't add Verify Role to ${member.user.tag}: ${e.message}`, 1);
                        return;
                    }

                    tools.log(scriptName, `${member.user.tag} has verified`);
                    return;
                }
            }
        });
    }

    function activate() {
        if (this.activated) return;

        tools.log(scriptName, `${scriptName} V${version} was reactivated`);
        this.activated = true;
    }

    function deactivate() {
        if (!this.activated) return;

        tools.log(scriptName, `${scriptName} V${version} was deactivated`);
        this.activated = false;
    }

    async function command(cmd, args) {
        if (!this.activated || !this.initialized) return;
        switch (cmd) {
            case "shutdown":
                tools.log(scriptName, `${scriptName} V${version} was deactivated`);
                this.activated = false;
                return true;
            case "reloaddb":
                this.db = JSON.parse(fs.readFileSync(`${dir}/db/discord-accept-rules-bot.db`, "utf-8"));
                tools.log(scriptName, `Database successfully reloaded`);
                return true;
            case "reloadconfig":
                this.config = JSON.parse(fs.readFileSync(`${dir}/config/discord-accept-rules-bot.json`, "utf-8"));
                tools.log(scriptName, `Config successfully reloaded`);
                return true;
            case "addverifymessage":
                if (!args[0]) {
                    tools.log(scriptName, `Please specify a channel ID to send the message to.`, 1);
                    return true;
                }

                let guild2 = client.guilds.get(this.config.guildid);
                if (!guild2) {
                    tools.log(scriptName, `The guild specified in the modules config was not found. Please check the config file. Verifymessage could not be sent.`, 1);
                    return true;
                }

                let channel = guild2.channels.get(args[0]);
                if (!channel) {
                    tools.log(scriptName, `The specified channel could not be found in the guild.`, 1);
                    return true;
                }

                try {
                    var embed = tools.makeEmbed(client.user, this.config.embedColor, this.config.verifyMessageTitle, this.config.verifyMessageDescription, this.config.verifyMessageFields);
                } catch (e) {
                    tools.log(scriptName, `Couldn't create Embed with config values: ${e.message}`, 1);
                    return true;
                }

                try {
                    var msg = await channel.send({
                        embed: embed
                    });
                } catch (e) {
                    tools.log(scriptName, `Couldn't send message to ${channel.name}: ${e.message}`, 1);
                    return true;
                }
                tools.log(scriptName, `Message was sent with ID ${msg.id}`);

                try {
                    await msg.react(this.config.verifyMessageReaction);
                } catch (e) {
                    tools.log(scriptName, `Couldn't add reaction to verify message in ${channel.name}: ${e.message}`, 1);
                    return true;
                }
                tools.log(scriptName, `Reaction was added`);

                this.db.messageid = msg.id;
                this.db.messagechannelid = msg.channel.id;
                fs.writeFileSync(`${dir}/db/discord-accept-rules-bot.db`, JSON.stringify(this.db));

                tools.log(scriptName, `Settings saved`);
                return true;
            case "verify":
                if (!args[0]) {
                    tools.log(scriptName, `Please provide the user ID of the person to verify`, 1);
                    return true;
                }

                let guild3 = client.guilds.get(this.config.guildid);
                if (!guild3) {
                    tools.log(scriptName, `The guild specified in the modules config was not found. Please check the config file.`, 1);
                    return true;
                }

                let member2 = guild3.members.get(args[0]);
                if (!member2) {
                    tools.log(scriptName, `The specified user was not found on the server.`, 1);
                    return true;
                }

                if (member2.roles.get(this.config.roleid)) {
                    tools.log(scriptName, `The specified user is already verified.`, 1);
                    return true;
                }

                let role = guild3.roles.get(this.config.roleid);
                if (!role) {
                    tools.log(scriptName, `The specified role was not found on the server. Please check modules config.`, 1);
                    return true;
                }

                try {
                    await member2.addRole(role);
                } catch (e) {
                    tools.log(scriptName, `Couldn't add Verify Role to ${member2.user.tag}: ${e.message}`, 1);
                    return true;
                }

                tools.log(scriptName, `${member2.user.tag} was manually verified.`);
                return true;
            case "unverify":
                if (!args[0]) {
                    tools.log(scriptName, `Please provide the user ID of the person to verify`, 1);
                    return true;
                }

                let guild4 = client.guilds.get(this.config.guildid);
                if (!guild4) {
                    tools.log(scriptName, `The guild specified in the modules config was not found. Please check the config file.`, 1);
                    return true;
                }

                let member3 = guild4.members.get(args[0]);
                if (!member3) {
                    tools.log(scriptName, `The specified user was not found on the server.`, 1);
                    return true;
                }

                if (!member3.roles.get(this.config.roleid)) {
                    tools.log(scriptName, `The specified user is not verified.`, 1);
                    return true;
                }

                let role2 = guild4.roles.get(this.config.roleid);
                if (!role2) {
                    tools.log(scriptName, `The specified role was not found on the server. Please check modules config.`, 1);
                    return true;
                }

                try {
                    await member3.removeRole(role2);
                } catch (e) {
                    tools.log(scriptName, `Couldn't remove Verify Role from ${member3.user.tag}: ${e.message}`, 1);
                    return true;
                }

                tools.log(scriptName, `${member3.user.tag} was manually unverified.`);
                return true;
            case "help":
                console.log(`===== Discord Accept Rules Module Commands\n\naddverifymessage <channelid>: Sends a verify message to the specified channel\nverify <DiscordID>:           Adds the specified role to the user\nunverify <DiscordID>:         Removes the specified role from the user\n`);
                return true;
            default:
                return;
        }
    }

    return {
        init,
        activate,
        deactivate,
        command
    }
}

main.js
/**
 * Discord Bot Mainframe
 * @author AstroGD - https://www.astrogd.eu
 * @since 2018-07-15
 */

const scriptName = 'main.js';

//Required Modules and Files
const Discord = require("discord.js");
const config = require(`${__dirname}/config/config.json`);
const fs = require("fs-extra");
const readline = require("readline");
const package = require(`${__dirname}/package.json`);

//Instances
const client = new Discord.Client();
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});
rl.setPrompt("> ");

//Required Scripts
const tools = require(`${__dirname}/tools/tools.js`)(Discord, client, fs, __dirname);

//Module Definition
const discord_accept_rules_bot = require(`${__dirname}/modules/discord-accept-rules-bot.js`)(client, fs, tools, __dirname);

//Module Commands
async function command(cmd, args) {
    let success = false;

    if (await tools.command(cmd, args)) success = true;
    if (await discord_accept_rules_bot.command(cmd, args)) success = true;

    return success;
}

//Module Initialization
async function initialization() {
    await discord_accept_rules_bot.init();
}


//DO NOT CHANGE THE CODE BELOW IF YOU DON'T KNOW WHAT YOU ARE DOING.
// 
// 
//CHANGING THE CODE COULD LEAD TO SECURITY ISSUES AND MALFUNCTION. DO NOT EVEN CHANGE SOMETHING IF SAID BY SOMEONE ELSE CLAIMING TO KNOW WHAT HE IS DOING!
//ONLY USE THE ORIGINAL DISCORD BOT MAINFRAME VERSION FROM ASTROGDs GITHUB REPOSITORY.
// 
// 
//IF YOU HAVE ANY QUESTIONS, FEEL FREE TO CONTACT ASTROGD:
// Mail: support@astrogd.eu
// Twitter: @astrogd
// Instagram: @astrogd
// Discord: https://go.astrogd.eu/discord (AstroGD#3416)
// 
// 
//THE CODE BELOW MAY ONLY BE CHANGED TO UPDATE THE DISCORD BOT MAINFRAME. IF THERE IS AN UPDATE AVAILABLE, YOU WILL GET A MESSAGE IN THE CONSOLE.
//FOR MORE INFOS ON HOW TO UPDATE, CHECK ASTROGDs REPOSITORY



//Initialization
async function init() {
    try {
        await tools.init();
    } catch (e) {
        console.log(`An error occurred during the initialization: ${e.stack}`);
        process.exit();
        return;
    }
    tools.log(scriptName, "Scripts initialized - Getting version information");

    global.VERSIONINFO = await tools.checkVersion();
    if (!VERSIONINFO) {
        tools.log(scriptName, "Version couldn't be fetched. Make sure you have a connection to the internet and access to software.astrogd.eu", 1);
    } else {
        if (tools.isVersionLower(VERSIONINFO.version, package.version)) tools.log(scriptName, `Theres a new version available for Discord accept rules bot by AstroGD (${package.version} --> ${VERSIONINFO.version})`, 2);
    }

    tools.log(scriptName, "Logging in...");
    try {
        await client.login(config.token);
    } catch (e) {
        tools.log(scriptName, `Authentication failed: ${e.message}`);
        process.exit();
        return;
    }
}

async function checkVersion() {
    VERSIONINFO = await tools.checkVersion();
    if (!VERSIONINFO) {
        tools.log(scriptName, "Version couldn't be fetched. Make sure you have a connection to the internet and access to software.astrogd.eu", 1);
        return false;
    }

    let newVerAvailable = false;
    if (tools.isVersionLower(VERSIONINFO.version, package.version)) {
        tools.log(scriptName, `Theres a new version available for Discord accept rules bot by AstroGD (${package.version} --> ${VERSIONINFO.version})`, 2);
        newVerAvailable = true;
    }

    return newVerAvailable;
}

client.on(`ready`, async () => {
    tools.log(scriptName, `Setup completed. Connected to ${client.user.tag} (${client.user.id})`);
    try {
        await initialization();
    } catch (e) {
        tools.log(scriptName, `Initalization failed: ${e.stack}`, 1);
        client.destroy().then(() => {
            process.exit();
        });
        return;
    }
    tools.log(scriptName, "All modules were successfully initialized");

    setInterval(checkVersion, 1*60*60*1000);

    rl.prompt(true);
});

rl.on("line", async (input) => {
    rl.pause();
    let cmd = input.split(" ")[0].toLowerCase();
    let args = input.split(" ").slice(1);

    tools.logFile(scriptName, `Command recieved: ${input}`);

    let success = await command(cmd, args);

    switch (cmd) {
        case "shutdown":
            console.log(`Shutting down...`);

            client.destroy();
            process.exit();
        case "help":
            console.log(`===== Discord Bot Mainframe Commands\n\nshutdown:                     Deactivates all correctly installed Modules and shuts the bot down\ncheckversion:                 Checks if there is a new version available\nhelp:                         Displays all commands available from Modules or Mainframe\nreloaddb:                     Reloads Database. Use only if changes were made manually\nreloadconfig:                 Reloads Config. Use only if changes were made manually\n\n=====`);
            break;
        case "checkversion":
            if (!await checkVersion()) console.log("Everythings up to date!");
            break;
        default:
            if (!success) {
                console.log(`${cmd}: The input "${cmd}" was not being recognized as a command. Please check your spelling or type "help" for a list of possible commands. Also check, if module commands were correctly installed.`);
                tools.logFile(scriptName, `Unknown Command`);
            }
            break;
    }

    rl.prompt();

});

rl.pause();
console.log(`Starting ${scriptName} V${package.version}...`);
init();